src/Microsoft.Extensions.OData.Migration/ExtensionMethods.cs

```diff
@@ -0,0 +1,143 @@
+                case EdmPrimitiveTypeKind.DateTimeOffset:
+                case EdmPrimitiveTypeKind.Duration:
+                case EdmPrimitiveTypeKind.TimeOfDay:
+                    return new EdmTemporalTypeReference(type, isNullable);
+                case EdmPrimitiveTypeKind.Geography:
+                case EdmPrimitiveTypeKind.GeographyPoint:
+                case EdmPrimitiveTypeKind.GeographyLineString:
+                case EdmPrimitiveTypeKind.GeographyPolygon:
+                case EdmPrimitiveTypeKind.GeographyCollection:
+                case EdmPrimitiveTypeKind.GeographyMultiPolygon:
+                case EdmPrimitiveTypeKind.GeographyMultiLineString:
+                case EdmPrimitiveTypeKind.GeographyMultiPoint:
+                case EdmPrimitiveTypeKind.Geometry:
+                case EdmPrimitiveTypeKind.GeometryPoint:
+                case EdmPrimitiveTypeKind.GeometryLineString:
+                case EdmPrimitiveTypeKind.GeometryPolygon:
+                case EdmPrimitiveTypeKind.GeometryCollection:
+                case EdmPrimitiveTypeKind.GeometryMultiPolygon:
+                case EdmPrimitiveTypeKind.GeometryMultiLineString:
+                case EdmPrimitiveTypeKind.GeometryMultiPoint:
+                    return new EdmSpatialTypeReference(type, isNullable);
+                default:
+                    throw new NotImplementedException();
+            }
+        }*/
+
+
+        /// <summary>
+        /// Attempts multiple casts on Data.Edm.IEdmType to reach derived classes and extract full name.
+        /// Implementation taken (copy-pasted and primitiveType case commented) from OData V4 ExtensionMethods.
+        /// </summary>
+        /// <param name="type">OData V3 type to extract name of</param>
+        /// <returns>Full name of V3 type</returns>
+        public static string FullTypeName(this Data.Edm.IEdmType type)
+        {
+            // Taken from OData v4 EdmConstants (internal)
+            const string CollectionTypeFormat = "Collection" + "({0})";
+
+            // No corresponding public class EdmCoreModelPrimitiveType in V3?
+            // TODO: check with Sam Xu
+            /*var primitiveType = type as Data.Edm.Library.EdmValidCoreModelPrimitiveType; // inaccessible because private.
+            if (primitiveType != null)
+            {
+                return primitiveType.FullName;
+            }*/
+
+            var namedDefinition = type as Data.Edm.IEdmSchemaElement;
+            var collectionType = type as Data.Edm.IEdmCollectionType;
+            if (collectionType == null)
+            {
+                return namedDefinition != null ? FullName(namedDefinition) : null;
+            }
+
+            // Handle collection case.
+            namedDefinition = collectionType.ElementType.Definition as Data.Edm.IEdmSchemaElement;
+
+            return namedDefinition != null ? string.Format(CultureInfo.InvariantCulture, CollectionTypeFormat, FullName(namedDefinition)) : null;

```

Comment:
Is formatting to InvariantCulture needed here? If not, it can be simplifed to 
return nameDefinition != null ? $"Collection({namedDefinition.FullName()})" : null;

Reply:
I took this method from the main OData repository, so I think it's probably safe to keep it